### ISSUE TITLE
Resolve remote git references using provider agnostic method

### ISSUE THREAD

**Issue by manics** [jupyterhub/binderhub#843]
Follow up from https://discourse.jupyter.org/t/how-to-use-bitbucket-repositories-on-mybinder-org/1000/4

### Proposed change
- Resolve branch names using `git ls-remote` in the bare git provider https://github.com/jupyterhub/binderhub/blob/0b8f99af0a23a7828810b03f010edcefc9a46195/binderhub/repoproviders.py#L126-L160
- Potentially also replace provider specific API calls to resolve branch names

E.g.:
```
git ls-remote https://bitbucket.org/ueacomputervision/image-labelling-tool
f3ddb33e4839f8a0fe73c168993b405adc13daf0        HEAD
8116559e4030d1d4f557617780410cbca7896663        refs/heads/canvas
d50105bacbb74ca139c5bf38ef1657190a006929        refs/heads/group
f3ddb33e4839f8a0fe73c168993b405adc13daf0        refs/heads/master
0b92e511ce8e04f26131efddde9b4a2f5ef9e1e1        refs/heads/multi-user
bc53e6aa12b6d7988de514c8c92e29877be5cc61        refs/heads/separate-modules
a1b2c3bcc2e563afbd335013805d19c95d67224d        refs/heads/server-side-refactor
```
```
$ git ls-remote https://bitbucket.org/ueacomputervision/image-labelling-tool master
f3ddb33e4839f8a0fe73c168993b405adc13daf0        refs/heads/master
```

As indicated in the linked discussion currently the resolution of branch-names to sha1s is done using provider specific API calls, for anything unsupported users must figure out the commit hash themselves.

### Who would use this feature?
Anyone not using GitHub or GitLab.

**Comment by betatim** [comment]
You think you know a lot about Git and then something like this comes along. Every day is a school day! We should definitely add this to the "bare git repo provider". For GitHub, Gitlab.com and others where we have an API we can call I'd leave things as is. It works, it is cheap and we have a good caching setup.

For bare git repositories this will be a major usability improvement!

**Comment by betatim** [comment]
Closed by #895
